src/concatenate.jl

```diff
@@ -53,14 +57,20 @@ struct Concatenated{Interface,Dims,Args<:Tuple}
   end
 end
 
-function Concatenated(interface::Union{Nothing,AbstractInterface}, dims::Val, args::Tuple)
+function Concatenated(interface::Nothing, dims::Val, args::Tuple)
   return _Concatenated(interface, dims, args)
 end
+function Concatenated(interface::AbstractArrayInterface, dims::Val, args::Tuple)
+  N = cat_ndims(dims, args...)
+  return _Concatenated(typeof(interface)(Val(N)), dims, args)
+end
 function Concatenated(dims::Val, args::Tuple)
-  return Concatenated(interface(args...), dims, args)
+  N = cat_ndims(dims, args...)
+  return _Concatenated(typeof(interface(args...))(Val(N)), dims, args)
 end
 function Concatenated{Interface}(dims::Val, args::Tuple) where {Interface}
-  return Concatenated(Interface(), dims, args)
+  N = cat_ndims(dims, args...)
+  return _Concatenated(set_interface_ndims(Interface, Val(N)), dims, args)
```

Comment:
With this new call to `similar`, do you need to set the interface ndims already at this point? I guess this would be automatically determined from the `ax` later down the line, so I'm just wondering if there is any difference.

Reply:
That's true, when you call `similar`, the `ndims` in the interface will get overridden by the axes input to `similar`. I kept it in case there is some other use case where storing the `ndims` in the interface is useful, but I admit I don't have a particular use case in mind for `Concatenated`.

I suppose what this does is catch cases where a user specifies an interface but it actually has the wrong `ndims` for the concatenation expression, I guess that could be checked.

But also thinking about this particular constructor, it should change the `Interface` type since that mean this constructor doesn't satisfy `T(args...) isa T`, so I'll change this one.